PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx

```diff
@@ -0,0 +1,580 @@
+  this->CmdUpdateTransform->SetBlocking(this->CommandBlocking);
+  this->CmdUpdateTransform->SetDeviceID(this->CommandIGTLIODeviceID);
+
+  this->CmdSaveConfig->SetCommandTimeoutSec(this->CommandTimeoutSec);
+  this->CmdSaveConfig->SetCommandName("SaveConfig");
+  this->CmdSaveConfig->SetBlocking(this->CommandBlocking);
+  this->CmdSaveConfig->SetDeviceID(this->CommandIGTLIODeviceID);
+
+}
+
+//----------------------------------------------------------------------------
+vtkSlicerPlusRemoteLogic::~vtkSlicerPlusRemoteLogic()
+{
+  this->CmdGetCaptureDeviceIDs->RemoveAllObservers();
+  this->CmdGetReconstructorDeviceIDs->RemoveAllObservers();
+  this->CmdStartVolumeReconstruction->RemoveAllObservers();
+  this->CmdStopVolumeReconstruction->RemoveAllObservers();
+  this->CmdReconstructRecorded->RemoveAllObservers();
+  this->CmdStartRecording->RemoveAllObservers();
+  this->CmdStopRecording->RemoveAllObservers();
+  this->CmdGetVolumeReconstructionSnapshot->RemoveAllObservers();
+  this->CmdUpdateTransform->RemoveAllObservers();
+  this->CmdSaveConfig->RemoveAllObservers();
+}
+
+//----------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::PrintSelf(ostream& os, vtkIndent indent)
+{
+  this->Superclass::PrintSelf(os, indent);
+}
+
+//-----------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::RegisterNodes()
+{
+  if (!this->GetMRMLScene())
+  {
+    vtkErrorMacro("vtkSlicerPlusRemoteLogic::RegisterNodes failed: MRML scene is invalid");
+    return;
+  }
+
+  this->GetMRMLScene()->RegisterNodeClass(vtkSmartPointer<vtkMRMLPlusRemoteNode>::New());
+  this->GetMRMLScene()->RegisterNodeClass(vtkSmartPointer<vtkMRMLPlusServerLauncherRemoteNode>::New());
+}
+
+//---------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::SetMRMLSceneInternal(vtkMRMLScene * newScene)
+{
+  vtkNew<vtkIntArray> events;
+  events->InsertNextValue(vtkMRMLScene::NodeAddedEvent);
+  this->SetAndObserveMRMLSceneEventsInternal(newScene, events.GetPointer());
+}
+
+//---------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::OnMRMLSceneNodeAdded(vtkMRMLNode* node)
+{
+  if (!node)
+  {
+    return;
+  }
+}
+
+//----------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::CreateDefaultParameterSet()
+{
+  if (!this->GetMRMLScene())
+  {
+    return;
+  }
+
+  std::vector<vtkMRMLNode*> nodes = std::vector<vtkMRMLNode*>();
+  this->GetMRMLScene()->GetNodesByClass("vtkMRMLPlusRemoteNode", nodes);
+
+  if (nodes.empty())
+  {
+    this->GetMRMLScene()->AddNewNodeByClass("vtkMRMLPlusRemoteNode");
+  }
+}
+
+//---------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::GetCaptureDeviceIDs(vtkMRMLIGTLConnectorNode* connectorNode, vtkCallbackCommand* callback)
```

Comment:
I don't think multiple modules can call GetCaptureDeviceIDs independently, as there is only a single CmdGetCaptureDeviceIDs command. So, we can simplify things and also make it more robust:
- remove vtkCallbackCommand callback argument and just update an associated PlusRemote MRML node with the result; the node will invoke Modified event (but you may also invoke some additional custom modified event that modules can observe)
- if a command is already in progress then return with an error

This is applicable to all commands.

Reply:
That's a good idea. It also solves the problem of accessing the commands from Python.